src/community/features-templating/features-templating-ows/src/main/java/org/geoserver/featurestemplating/ows/TemplateCallback.java

```diff
@@ -191,23 +207,27 @@ private Response getTemplateFeatureInfoResponse(GetFeatureInfoRequest request) {
         String infoFormat = request.getInfoFormat();
         TemplateIdentifier identifier =
                 TemplateIdentifier.fromOutputFormat(request.getInfoFormat());
-        int nTemplates = 0;
-        for (MapLayerInfo li : layerInfos) {
-            if (li != null && li.getResource() instanceof FeatureTypeInfo) {
-                if (ensureTemplatesExist(li.getFeature(), identifier.getOutputFormat()) != null)
-                    nTemplates++;
-            }
-        }
         Response result = null;
-        if (nTemplates > 0) {
-            int size = layerInfos.size();
-            if (size != nTemplates)
-                throw new ServiceException(
-                        "To get a features templating getFeatureInfo a template is needed for every FeatureType but "
-                                + (size - nTemplates)
-                                + " among the requested ones are missing a template");
+        if (identifier != null) {
+            int nTemplates = 0;
+            for (MapLayerInfo li : layerInfos) {
+                if (li != null && li.getResource() instanceof FeatureTypeInfo) {
+                    if (ensureTemplatesExist(li.getFeature(), identifier.getOutputFormat()) != null)
+                        nTemplates++;
+                }
+            }
+            if (nTemplates > 0) {
+                int size = layerInfos.size();
+                if (size != nTemplates)
+                    throw new ServiceException(
+                            "To get a features templating getFeatureInfo a template is needed for every FeatureType but "
+                                    + (size - nTemplates)
+                                    + " among the requested ones are missing a template");
 
-            result = OWSResponseFactory.getInstance().featureInfoResponse(identifier, infoFormat);
+                result =
+                        OWSResponseFactory.getInstance()
+                                .featureInfoResponse(identifier, infoFormat);
+            }
```

Comment:
This method is quite complicated and hard to read, I would suggest the following re-writing:
```java
   /**
     * Helper method to find the response for a WMS GetFeatureInfo request.
     * <p/>
     * If none of the requested layers uses a feature template for the requested info format, then we delegate on
     * whatever is the default response for the requested info format.
     * <p/>
     * If at least one of the requested layers uses a feature template for the requested info format, then all the
     * requested layers need to have a matching feature template for the requested info format, otherwise an
     * exception will be thrown.
     */
    private Response getTemplateFeatureInfoResponse(GetFeatureInfoRequest request) {
        // let's check if feature templating supports the requested info format
        String infoFormat = request.getInfoFormat();
        TemplateIdentifier identifier =
                TemplateIdentifier.fromOutputFormat(request.getInfoFormat());
        if (identifier == null) {
            // features templating doesn't support the requested info type,
            // hence we delegate on whatever is the default response
            return null;
        }
        // the request info format is supported by features templating
        // let's check if the requested layers all have a template supporting
        // the requested info format
        List<MapLayerInfo> layers = request.getQueryLayers();
        int matchingTemplates = 0;
        for (MapLayerInfo layer : layers) {
            if (layer != null
                    && layer.getResource() instanceof FeatureTypeInfo
                    && (ensureTemplatesExist(layer.getFeature(), identifier.getOutputFormat()) != null)) {
                // there is an associated template with the layer matching the requested output format
                matchingTemplates++;
            }
        }
        if (matchingTemplates == 0) {
            // no layers is using features templating, hence we delegate in whatever is the default info format
            return null;
        }
        // if one of the requested layer uses a features template, then all the other need to have one too
        int matchTemplatesDiff = layers.size() - matchingTemplates;
        if (matchTemplatesDiff != 0) {
            // there is at least one requested layer that doesn't have template matching the requested info format
            // this situation is currently not supported
            throw new ServiceException(
                    "In a WMS GetFeatureInfo all requested layers need to have a template matching " +
                            "the expected info format, " + matchTemplatesDiff +
                            " among the requested ones are missing such template.");
        }
        // we are good, all the requested layers have a template matching the requested info format
        return OWSResponseFactory.getInstance().featureInfoResponse(identifier, infoFormat);
    }
```

Reply:
fixed